Adding custom exception for reclassification error

dcdenu4 opened the pull request:
Adding a custom exception to `pygeoprocessing/geoprocessing.py` called `ReclassificationMissingValuesError` that takes a _message_, _missing values_, and _value map_. These are set as attributes that can then be accessed via InVEST for example.

Added the importing of this custom class to `__init__.py` so it is available via `pygeoprocessing.ReclassificationMissingValuesError`.

Closes #105

dcdenu4 committed:
I. #105 adding custom exception for reclassify

Adding the importing of the custom exception in init so that it is
accessible via pygeoprocessing.ReclassificationMissingValuesError

dcdenu4 committed:
I. #105 fixing test

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -113,13 +116,11 @@ def test_reclassify_raster_missing_pixel_value(self):
         }
         target_nodata = -1
-        with self.assertRaises(ValueError) as cm:
+        with self.assertRaises(reclass_error) as cm:
```
In combination with line 102 above, I'd tend to do an `from .geoprocessing import ReclassificationMissingValuesError` then `self.assertRaises(ReclassificationMissingValuesError)` rather than obscure the exception through a variable. Or just `pygeoprocessing.ReclassificationMissingValuesError`. It's one of the reasons I kinda like custom exceptions when they make sense -- you can infer their meaning from their class name.

dcdenu4 replied:
Yep, agreed. This was me trying to appease the `line too long` linting gnomes

dcdenu4 replied:
Now using `pygeoprocessing.ReclassificationMissingValuesError` straight up in https://github.com/natcap/pygeoprocessing/pull/106/commits/8509356492a7b45beaa058e8184e09d1087603a7

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -113,13 +116,11 @@ def test_reclassify_raster_missing_pixel_value(self):
-        expected_message = (
-            'The following 1 raster values [-0.5] from "%s" do not have ' %
-            (raster_path,))
+        expected_message = ('The following 1 raster values [-0.5]')
```
I think you can lose the parens here now that it's on one line.

dcdenu4 replied:
Parens removed in https://github.com/natcap/pygeoprocessing/pull/106/commits/8509356492a7b45beaa058e8184e09d1087603a7

dcdenu4 committed:
I. #105 removing blank space in imports

richpsharp reviewed (changes requested):
Easy Doug! I had a couple stylistic comments, but also can you add an entry in `HISTORY.rst` since this is a user-facing change? I think a brief statement of guidance of the expected use of that exception would be helpful there too.

dcdenu4 committed:
I. #105 exposing exception name

use the custom exception name and don't hide it with a variable

phargogh commented on src/pygeoprocessing/__init__.py:
```diff
@@ -7,6 +7,7 @@
 import types
 
 from . import geoprocessing
+from .geoprocessing import ReclassificationMissingValuesError
```
Could you add `ReclassificationMissingValuesError`  to `__all__` as well here?  Doing so will enable `from pygeoprocessing import *` to also expose `ReclassificationMissingValueError`.

dcdenu4 replied:
Added to `__all__` here: https://github.com/natcap/pygeoprocessing/pull/106/commits/5eecc74c5c71675b465c97261efedacbe8ce3f8e

phargogh reviewed (changes requested):
Looks like a good solution to me!  I had one comment about including this in `__all__`, and then since this is a user-facing change, could you include a note in `HISTORY` about this?  Thanks!

richpsharp reviewed (changes requested):
Thanks for making those changes Doug. Did you see my note about `HISTORY.rst`? Can you update that to show the user facing changes and a description about how this exception could occur and why a user may want to catch this exception?

dcdenu4 committed:
I. #105 adding custom exception to __all__

dcdenu4 committed:
I. #105 adding description to History for exception

dcdenu4 committed:
I. #105 updating docstring and adding attribute

Adding attribute ``msg`` to docstring and updating ``reclassify_raster``
docstring for the custom exception being raised

dcdenu4 commented:
@richpsharp @phargogh I took a stab at a useful History blurb here: https://github.com/natcap/pygeoprocessing/pull/106/commits/da813ea879abf03582c072a77c29b1946dcca299

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -29,6 +29,23 @@
+        msg (str) - error message
+        missing_values (list) - a list of the missing values from the raster
+            that are not keys in the dictionary
+        value_map (dict) - the dictionary used to map raster values
```
Hey Doug, I'm sorry I didn't notice this before until you made a note of it in your HISTORY.rst. Presumably the caller is in the same context as the exception that's raised here so the `value_map` should be in scope of handling that exception right? The exception should be providing information the caller does not have, which are the actual missing values. It's not necessarily bad to include this, but naively it seems unnecessary and just another bit of code that could be dropped. Is there a reason you're including `value_map` as an attribute in the exception?

dcdenu4 replied:
Nope, that's a fair assessment and does seem redundant. I'll remove it.

dcdenu4 replied:
Removed `value_map` here: https://github.com/natcap/pygeoprocessing/pull/106/commits/0ed7d528c4fe110c9333a5877e5d9bc00b902482

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1718,8 +1735,9 @@ def reclassify_raster(
-           ``base_raster_path_band`` is not a key in ``attr_dict``.
+        ReclassificationMissingValuesError if ``values_required`` is ``True``
+        and a pixel value from ``base_raster_path_band`` is not a key in
+        ``attr_dict``.
```
Sorry, I didn't notice tihs earlier either, but I think ``attr_dict`` should be ``value_map``. I'm guessing that variable changed a while ago and we didn't change the docstring.

dcdenu4 replied:
Good catch.

dcdenu4 replied:
Removed `value_map` here: https://github.com/natcap/pygeoprocessing/pull/106/commits/0ed7d528c4fe110c9333a5877e5d9bc00b902482

richpsharp commented on HISTORY.rst:
```diff
@@ -3,6 +3,15 @@ Release History
+  * ``value_map`` - the dictionary the raster values were attempted to map to
+  This error is now raised in ``reclassify_raster`` and can be caught 
+  to better debug the mismatch when using a table to reclassify a raster. 
 * Correcting the docstring for ``pygeoprocessing.numpy_array_to_raster`` to
```
I've got a proposed re-word of this section below. The ``msg`` is part of the base exception so I don't think that needs mentioning and if you're up for my suggestion below I think you can drop ``value_map``. 

`Added a custom exception ``ReclassificationMissingValuesError`` to ``pygeoprocessing.reclassify_raster`` that raises when a raster pixel value is not represented in ``value_map``. This custom exception provides a list of missing raster pixel values in a ``missing_values`` attribute that allows the caller access to the pixel values that are missing through a Python type rather than indirectly through an error message.`

dcdenu4 replied:
Yea, I think this is better stated then what I have. I'll update it.

dcdenu4 replied:
Updated statement here: https://github.com/natcap/pygeoprocessing/pull/106/commits/d0bcb5436d271f8f1c58a6d44853cdbd782fb185

richpsharp reviewed (changes requested):
Sorry I have more review questions here Doug. I didn't look at this as carefully as I should have on a first pass.

dcdenu4 commented:
Thanks for the comments Rich, I also noticed a test is failing because of the `__all__` inclusion. Looks like that test excepts all the attributes to be functions, which this is not.

dcdenu4 committed:
I. #105 adding MethodType for __all__ check

dcdenu4 committed:
I. #105 updating History blurb

dcdenu4 committed:
I. #105 remove value_map

also fix ``attr_dict`` to ``value_map`` in docstring

dcdenu4 committed:
I. #105 testing exception agains class

For the star import test, adding an or statement to check if attribute
is a subclass of Exception class, which covers custom exceptions we
might have.

phargogh reviewed (approved):
👍
